### ISSUE TITLE
[FEATURE] Data Line Width

### ISSUE THREAD

**Issue by dg-pb** [bokeh/bokeh#13494]
### Problem description

This is just a follow up on [stack](https://stackoverflow.com/questions/60411960/using-data-line-widths-in-bokeh)

The owner of the stack thread decided not to issue a feature request, but I decided to do so.

### Feature description

Simply allowing specification of line width in data coordinates.

I had similar issue (although not exactly the same) with circles, where circles were not stretched according to data coordinates, but fortunately there was already an alternative - `ellipse`, which did exactly what I needed.

However, I can not seem to find anything simple for lines, except for...

### Potential alternatives

drawing polygons manually. However, this is a fairly involved process for operation, which is very basic for many vector libraries. In this instance, I was using `kiva`, which wraps many other libraries and line width seems to be in data coordinates by default in all of those.

So I think having this, would make a more seamless transition for those navigating from those libraries towards bokeh.

### Additional information

Thank you.

**Comment by dgrigonis** [comment]
Just to extend. Polygon doesn't work for me properly.

This is a collection of lines drawn in a polygon:
![](https://pasteboard.co/YdtVkoQU9mqJ.png)

And this is darkest version of it zoomed in:
![](https://gcdnb.pbrd.co/images/1DRc9ngt3MdC.png?o=1)

**Comment by bryevdv** [comment]
The underlying canvas implementation does not support anything like that -- at the lowest level it's only possible to draw lines on the canvas by providing pixel widths. So in principle we'd have to take into account (and record, and keep track of) the relative scale between the user-supplied "data width" and the original axis data scale, and change it automatically during any zoom, etc. 

But that is not even the actual main or big problem, which is that in general, that relative scale between axis range and date line width is *not well-defined*. Specifically, there is nothing to guarantee that the x-scale and y-scale for data-to-pixel transforms agree, so how do you even measure distance in order to know what you need to change on a line running diagonally? Do you just arbitrarily pick one axis or the other to measure with? Try to take some kind of complicated average? How is this supposed to work during a *single-axis* zoom? 

That's not even getting into categorical scales where a "data" line width setting doesn't really have any meaning at all, or a log scale where technically speaking it would have to change continuously in a very complicated way along the entire length of the line. But `Line` still would have the the "data" line width property everywhere (unless we rip up the curent api in a backward incompatible way and have different versions of "line" for different scales), so we'd have to explain all these special "it works, *except in these cases...*" caveats throughout the docs, which is never a good sign.

I don't mean to be overly down out of the gate but I do think that this is something that would add a large amount of complexity, and even in the best case still only be able to work out of the box "correctly" in a few specialized cases (i.e. with aspect matching enabled) and therefore be a magnet for documentation issues and support burden. So I am pretty 👎 on even attempting it. 


cc @bokeh/dev

**Comment by dg-pb** [comment]
I kinda achieved it nicely using polygons with fill and setting line alpha to 0. Maybe there is no need to actually invent anything from scratch, but instead make an abstraction for such usage. e.g. `p.multi_polygon_lines()` with interface such as `multi_lines`, but draws polygons instead.

After all, line with width is a polygon. So maybe having polygon abstractions, where this functionality is missing could be useful.

Obviously this depends on how desirable it is, but given that many vector graphic libraries work this way by default, I would guess it is of a fair benefit, given minimal amount of work if it is a simple abstraction.

**Comment by bryevdv** [comment]
> but given that vector graphic libraries works this way by default

Bokeh is not a vector graphics library, it is a raster canvas library, unavoidably dealing in pixels at the lowest level we target, which is the fundamental difference.

But the difficult question remains: what does "data line width" even mean when the x-range is (-10e7, 10e7) and the y-range is (0, 1) but the number of pixels is roughly the same in both dimensions?

**Comment by dg-pb** [comment]
Even though things are so under the hood, I found that bokeh is good enough replacement for vector graphic library.

To be specific, I model in pixel space, then need to check the result to be as close to what it would look in real life. Plotting pixel image doesn't do the courtesy, but vector graphics provide rather good indication. I have just started to test bokeh, and it provides result similar to that of vector graphics (plus all of the added benefits of bokeh).

And in all fairness, vector graphics also deal in pixels (at least eventually). The tangible effect from user's perspective is that `the model` is stored somewhere and is re-rendered on certain actions. Whether this is done by the library itself or somewhere lower, doesn't make much difference to me as long as I get the benefit.

Edit: I might have sounded too confident here. I am not an expert in graphics libraries and am just providing a viewpoint from user's perspective.

**Comment by dg-pb** [comment]
> But the difficult question remains: what does "data line width" even mean when the x-range is (-10e7, 10e7) and the y-range is (0, 1) but the number of pixels is roughly the same in both dimensions?

You just calculate it from x and y ranges. Calculate unit normals to draw 2 parallel lines at a distance of half width and draw a polygon. Then let polygon drawing take care of pixels. Are there any issues with the example you provided that I am missing?

**Comment by dg-pb** [comment]
This is what I have done myself:
```python
    @staticmethod
    def find_parallel(x, y, d):
        x_n, y_n = x[1] - x[0], y[1] - y[0]
        dist = (y_n**2 + x_n**2)**0.5
        ux, uy = y_n / dist, -x_n / dist
        return (x[0] + ux * d, x[1] + ux * d), (y[0] + uy * d, y[1] + uy * d)

    @classmethod
    def make_poly_lines(cls, xs, ys, width):
        """Fixed data width line input conversion
            from:   p.multi_line input
            to:     p.multi_polygons
        """
        xs_poly = list()
        ys_poly = list()
        for xl, yl in zip(xs, ys):
            for x, y in zip(mit.windowed(xl, 2), mit.windowed(yl, 2)):
                xd, yd = find_parallel(x, y, -width / 2)
                xu, yu = find_parallel(x, y, width / 2)
                xs_poly.append([[xd + (xu[1], xu[0])]])
                ys_poly.append([[yd + (yu[1], yu[0])]])
        return xs_poly, ys_poly
```

It is enough for my use case, however it could be done nicely in bokeh, where this conversion is done before sending to json, so that ColumnDataSource looks exactly like the one would do for `p.multi_line`.

Also, joining polygons without gaps could be done as it currently looks like this:

![](https://gcdnb.pbrd.co/images/6GDg5LxkS4bK.png?o=1)

**Comment by bryevdv** [comment]
@bokeh/core I am inclined to close this with `noaction` since I judge that it would be a fairly significant amount of complexity to add a glyph that does this, for a rather niche usage scenario. I also would not really support a Python API to "fake it" as above, since the results (i.e. without joins) are not really of acceptable quality and it would be fairly inefficient protocol-wise as well.

**Comment by mattpap** [comment]
Redefining the semantics of `line_width` is a huge no-go. However, we could add line-like glyph(s), i.e. fat lines and maybe fat arrows. We could potentially reuse line drawing algorithm from our webgl backend for this (to get joints' geometries right).

**Comment by bryevdv** [comment]
I agree that the only possible approach would be an entirely new glyph built to cater to this specific usage. Question is whether we think realistically anyone will take this up.

**Comment by ianthomas23** [comment]
> Redefining the semantics of `line_width` is a huge no-go. However, we could add line-like glyph(s), i.e. fat lines and maybe fat arrows. We could potentially reuse line drawing algorithm from our webgl backend for this (to get joints' geometries right).

:+1: on this.

**Comment by dg-pb** [comment]
> @bokeh/core I am inclined to close this with `noaction` since I judge that it would be a fairly significant amount of complexity to add a glyph that does this, for a rather niche usage scenario. I also would not really support a Python API to "fake it" as above, since the results (i.e. without joins) are not really of acceptable quality and it would be fairly inefficient protocol-wise as well.

Just to point out - adding joints would not be a big issue. As I was experimenting, everything was in data coordinates so it is a simple problem. I just didn't get down to it as I needed relatively thin lines, where joints didn't matter much.

I am currently on hold with that project, but will come back to it in due time and will experiment some more, or hopefully will try out new implementation and will have no need to do hacking.
